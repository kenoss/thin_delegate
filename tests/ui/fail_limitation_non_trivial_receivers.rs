// `#[thin_delegate::fill_delegate]` doesn't correctly fill trait methods with receivers except for
// `&self`, `&mut self`, `self`, as delegating to sutch types are not trivial.
//
// For example, consider to define `hello_box(self: Box<Self>)` for a type `struct Hoge(String)`.
// `Hoge::hello_box()` receives `Box<Hoge>`, but the inner type is not boxed.

use std::pin::Pin;
use std::rc::Rc;
use std::sync::Arc;

#[thin_delegate::register]
pub trait Hello {
    fn hello_ref(&self) -> String;
    fn hello_ref_mut(&mut self) -> String;
    fn hello_consume(self) -> String;
    fn hello_box(self: Box<Self>) -> String;
    fn hello_rc(self: Rc<Self>) -> String;
    fn hello_arc(self: Arc<Self>) -> String;
    fn hello_pin(self: Pin<&Self>) -> String;
}

impl Hello for String {
    fn hello_ref(&self) -> String {
        format!("hello, {self}")
    }

    fn hello_ref_mut(&mut self) -> String {
        format!("hello, {self}")
    }

    fn hello_consume(self) -> String {
        format!("hello, {self}")
    }

    fn hello_box(self: Box<Self>) -> String {
        format!("hello, {self}")
    }

    fn hello_rc(self: Rc<Self>) -> String {
        format!("hello, {self}")
    }

    fn hello_arc(self: Arc<Self>) -> String {
        format!("hello, {self}")
    }

    fn hello_pin(self: Pin<&Self>) -> String {
        format!("hello, {self}")
    }
}

#[thin_delegate::register]
struct HogeStruct(String);

#[thin_delegate::fill_delegate]
impl Hello for HogeStruct {}

fn main() {}
