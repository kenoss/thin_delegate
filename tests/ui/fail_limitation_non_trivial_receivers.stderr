error[E0308]: mismatched types
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:55:1
   |
16 |     fn hello_box(self: Box<Self>) -> String;
   |        --------- arguments to this function are incorrect
...
55 | #[thin_delegate::fill_delegate]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Box<_>`, found `String`
   |
   = note: expected struct `Box<_>`
              found struct `String`
   = note: for more on the distinction between the stack and the heap, read https://doc.rust-lang.org/book/ch15-01-box.html, https://doc.rust-lang.org/rust-by-example/std/box.html, and https://doc.rust-lang.org/std/boxed/index.html
note: method defined here
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:16:8
   |
16 |     fn hello_box(self: Box<Self>) -> String;
   |        ^^^^^^^^^
   = note: this error originates in the attribute macro `::thin_delegate::__internal__fill_delegate` which comes from the expansion of the attribute macro `thin_delegate::fill_delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: store this in the heap by calling `Box::new`
   |
55 | Box::new(#[thin_delegate::fill_delegate])
   | +++++++++                               +

error[E0308]: mismatched types
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:55:1
   |
17 |     fn hello_rc(self: Rc<Self>) -> String;
   |        -------- arguments to this function are incorrect
...
55 | #[thin_delegate::fill_delegate]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Rc<_>`, found `String`
   |
   = note: expected struct `Rc<_>`
              found struct `String`
note: method defined here
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:17:8
   |
17 |     fn hello_rc(self: Rc<Self>) -> String;
   |        ^^^^^^^^
   = note: this error originates in the attribute macro `::thin_delegate::__internal__fill_delegate` which comes from the expansion of the attribute macro `thin_delegate::fill_delegate` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:55:1
   |
18 |     fn hello_arc(self: Arc<Self>) -> String;
   |        --------- arguments to this function are incorrect
...
55 | #[thin_delegate::fill_delegate]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Arc<_>`, found `String`
   |
   = note: expected struct `Arc<_>`
              found struct `String`
note: method defined here
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:18:8
   |
18 |     fn hello_arc(self: Arc<Self>) -> String;
   |        ^^^^^^^^^
   = note: this error originates in the attribute macro `::thin_delegate::__internal__fill_delegate` which comes from the expansion of the attribute macro `thin_delegate::fill_delegate` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:55:1
   |
19 |     fn hello_pin(self: Pin<&Self>) -> String;
   |        --------- arguments to this function are incorrect
...
55 | #[thin_delegate::fill_delegate]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Pin<&_>`, found `String`
   |
   = note: expected struct `Pin<&_>`
              found struct `String`
note: method defined here
  --> tests/ui/fail_limitation_non_trivial_receivers.rs:19:8
   |
19 |     fn hello_pin(self: Pin<&Self>) -> String;
   |        ^^^^^^^^^
   = note: this error originates in the attribute macro `::thin_delegate::__internal__fill_delegate` which comes from the expansion of the attribute macro `thin_delegate::fill_delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
